Extends the functional paradigm section.

grakic opened the pull request:
State the key elements of functional paradigm supported by PHP and extends the example.

grakic committed:
Extends the functional paradigm section.

State the key elements of functional paradigm supported by PHP and extends the example.

wilmoore commented on _posts/03-02-01-Programming-Paradigms.md:
```diff
@@ -15,19 +15,68 @@ PHP 5.4 (2012).
-PHP has had annonymous functions since PHP 5.3:
+PHP supports first-class functions. It is possible to define a new function and assign it to a variable name and built-in functions
+can be referenced and called dynamically. Functions can be passed as arguments to other functions (Higher-order functions) and function
+can return other functions. New anonymous functions (also known as Closures) are present since PHP 5.3 (2009).
```
anonymous functions !== Closures (however, closures are anonymous functions)

grakic replied:
Agree with that. I did not want to argue with both the PHP manual and the RFC :)

On the etymology, once bound to the environment (e.g. closed) is Closure still a Closure or it becomes an Anonymous function?

wilmoore commented on _posts/03-02-01-Programming-Paradigms.md:
```diff
@@ -15,19 +15,68 @@ PHP 5.4 (2012).
+/**
+ * Takes two single variable functions f and g, and creates a new function f∘g
+ */
+function combine($f, $g)
```
This example fine; however, it could stand to be less academic and focus more on how one might solve a real-world problem with an anonymous function. Further, it would be good to separate the examples so there is one for a plain anonymous function and one that depicts a use-case where a closure would be in order.

wilmoore commented on _posts/03-02-01-Programming-Paradigms.md:
```diff
@@ -15,19 +15,68 @@ PHP 5.4 (2012).
+/**
+ * Creates an anonymous criteria function accepting items > $min
+ */
+function criteria_greater_than($min)
```
This seems like a good example. Perhaps the earlier examples could be part of the same theme and each could incrementally build upon the previous. That would give a new reader an easy walk through of how each relates and where/when to use each paradigm.

grakic committed:
anonymous functions !== closures

grakic committed:
Merge some changes from @wilmoore given in issue #72

grakic commented:
@philsturgeon @wilmoore It would be great to have more comments on the suggested changes so we can iterate this until it is ready for the merge.

grakic committed:
Improve functional programming section

- Change criteria for filter
- Extend the first example to show anonymous function without assignment
- Add paragraph on recursion
- Wrap to 120 chars

grakic committed:
Closures and variable scope

codeguy commented:
Same as I said in #72, even this is a bit longer than what I want on the main page. Keep a brief/basic description of each on the home page, link to a subpage in `pages/`. Subpage will use the default layout.

Both PRs seem to have similar content. Do you two want to agree on one PR to move forward with?

wilmoore commented:
@codeguy I'm happy enough with either and @grakic is showing a stronger preference for #67 so I'm happy to go that way. No worries here. I tried to incorporate both into one; however, my aggregation didn't seem to make it into his second revision and I am not sure I want to spend the time trying to integrate a second time.

grakic commented:
Can we go with just the first three paragraphs? @wilmoore, what do you think?

> PHP supports first-class function, meaning that a function can be assigned to a variable. Both user defined and built-in  functions can be referenced by a variable and invoked dynamically. Functions can be passed as arguments to other functions (feature called Higher-order functions) and function can return other functions.
> 
> Recursion, a feature that allows a function to call itself is supported by the language, but most of the PHP code will focus on iteration.
> 
> New anonymous functions (with support for closures) are present since PHP 5.3 (2009).

Currently this section is not very good. Example is misleading (is greet just the name of a function?). Also functional programming is not just about anonymous functions, and others features supported by PHP should be given some space.

codeguy commented:
@grakic @wilmoore Let's move forward with this PR. I'm closing the other. @grakic I'm not saying the existing text on the main page is perfect. Feel free to edit it, but keep it brief... link to a standalone page that includes the full text that you guys have created.

wilmoore commented:
> Can we go with just the first three paragraphs? @wilmoore, what do you think?

First three seems fair.

codeguy commented:
Agreed.

grakic commented:
Great. I will create a new directory for pages, move the content over there and include first 3 paragraphs on the main page. It  will be ready in a short time.

codeguy commented:
Awesome. Thanks guys.

grakic commented:
We need a different layout for topic pages. Created issue #110 to discuss.

grakic committed:
Move to a separate topic page

grakic committed:
Add page title

grakic commented:
Rebased on new content pages, keeping pull request history.

FYI, I will be mostly away for the next two weeks.

grakic commented:
@codeguy Is there something wrong with this pull request? Am I missing something?

If you do not want this content in the document it would be nice to say it so.

codeguy commented:
@grakic No. I've been super busy moving into a new home, getting utils setup, etc. I'll take a look and get pulled in later today.

grakic committed:
Copy new paragraph by @jeremeamia from main page to content page

grakic commented:
Rebased one more time on top of changes from #122. This is becoming a little frustrating.

philsturgeon commented:
Done, thanks for keeping at it!
